### PR TITLE
Кочетов - Лабораторная работа #1

### PR DESCRIPTION
**Pull request by quakerfisher**
Created lab-1 (CreditCalculator)

### CODE REVIEW

**Comment by quakerfisher** [comment]
@qwert183, @DariaKolistratova Leave a review, please

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+
+package ru.unn.agile.Credit.Model;
+
+import java.security.InvalidParameterException;
+
+public class Credit {
+    private final double s;
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: Please, change to "sum", s makes no sence

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    private final double m;
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: month

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    private final double p;
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: percents

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    Credit(final double s, final double m, final double p){
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: space before {

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        if(s < 0){
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: space

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            throw new InvalidParameterException("Sum must be positive!");
+        }
+        if(m < 0){
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: space

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            throw new InvalidParameterException("Months must be positive!");
+        }
+        if((p < 0) || (p > 100)){
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: space

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+            throw new InvalidParameterException("Percents must be from 0 to 100!");
+        }
+        if(m == 0) {
+            throw new InvalidParameterException("Months must be not null!");
+        }
+        if((s == 0) && (m == 0) && (p == 0)){
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: space

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        double normalized_p = (p / 100) / 12;
+        return s * (normalized_p + (normalized_p / (Math.pow(1 + normalized_p, m) - 1)));
+    }
+
+    public double countTotalSum() {
+        double a = countPayment();
+        return a * m;
+    }
+
+    public double countOverPayment() {
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: maybe countOverpayment, as I remember it is written in one word

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        if((s == 0) && (m == 0) && (p == 0)){
+            throw new InvalidParameterException("Parameters are wrong!");
+        }
+        this.s = s;
+        this.m = m;
+        this.p = p;
+    }
+
+    public double countPayment() {
+        double normalized_p = (p / 100) / 12;
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: magic numbers

File: code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+import java.security.InvalidParameterException;
+
+import static org.junit.Assert.*;
+
+public class CreditTest {
+    private final double delta = 0.4;
+
+    public Credit inputValues(){
+        double s = 10000;
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java]: Maybe will be better to make constants or test class fields and reuse later. (when we need negative, we multiply constant with -1)

Comment (DariaKolistratova) [inline code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java]: And we'll need no inputValues method, code will be simpler

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: Please remove space

File: code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
+
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java]: Please remove space

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+            throw new InvalidParameterException("Parameters are wrong!");
+        }
+        this.s = s;
+        this.m = m;
+        this.p = p;
+    }
+
+    public double countPayment() {
+        double normalized_p = (p / 100) / 12;
+        return s * (normalized_p + (normalized_p / (Math.pow(1 + normalized_p, m) - 1)));
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: The line seems too long

File: code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Test
+    public void canCreateCredit() {
+        Credit cred = inputValues();
+        assertNotNull(cred);
+    }
+
+    @Test(expected = InvalidParameterException.class)
+    public void canCountNegativeSum(){
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java]: space before {

File: code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+        double s = -10000;
+        double m = 12;
+        double p = 20;
+        Credit cred = new Credit(s, m, p);
+    }
+
+    @Test(expected = InvalidParameterException.class)
+    public void canCountNegativeMonths(){
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java]: space before {

File: code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+        double s = 10000;
+        double m = -12;
+        double p = 20;
+        Credit cred = new Credit(s, m, p);
+    }
+
+    @Test(expected = InvalidParameterException.class)
+    public void canCountWrongPercents(){
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java]: space before {

File: code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        double s = 10000;
+        double m = 12;
+        double p = -20;
+        Credit cred = new Credit(s, m, p);
+    }
+
+    @Test(expected = InvalidParameterException.class)
+    public void canCountCreditWhenMonthsAreNull(){
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java]: space before {

**Comment by quakerfisher** [comment]
@qwert183, @DariaKolistratova Leave a review, please

**Comment by qwert183** [comment]
Credit.java
1. months>0 and months=0 unite together
2. add if persent=0
3. delete if (s&m&p=0)

CreditTEst.java
1.  In  "public void canCountCreditWhenParametersAreUndefined()"
rename count -> create
and in some other functions.

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+
+    Credit(final double sum, final double months, final double percent) {
+        if (sum < 0) {
+            throw new InvalidParameterException("Sum must be positive!");
+        }
+        if (months < 0) {
+            throw new InvalidParameterException("Months must be positive!");
+        }
+        if (percent < 0 || percent > PERCENT_MAX_VALUE) {
+            throw new InvalidParameterException("Percents must be from 0 to 100!");
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: too long line

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+        if (sum == 0 && months == 0 && percent == 0) {
+            throw new InvalidParameterException("Parameters are wrong!");
+        }
+        this.sum = sum;
+        this.months = months;
+        this.percent = percent;
+    }
+
+    public double countPayment() {
+        double normalPercent = (percent / PERCENT_MAX_VALUE) / MONTHS_IN_THE_YEAR;
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: too long line

File: code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+        double fractionCountPay = normalPercent / (Math.pow(1 + normalPercent, months) - 1);
```
Comment (DariaKolistratova) [inline code/kochetov-pavel/src/main/java/ru/unn/agile/Credit/Model/Credit.java]: too long line

**Comment by DariaKolistratova** [comment]
👍

**Comment by quakerfisher** [comment]
@kirill-korniakov

File: code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.security.InvalidParameterException;
+
+import static org.junit.Assert.*;
+
+public class CreditTest {
+    private final double delta = 0.4;
+    private final double sum = 10000;
+    private final double months = 12;
+    private final double percent = 20;
```
Comment (Nikita94) [inline code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java]: если это константа, то они должны быть так же static и название должно быть оформлено  большими буквами. Я бы не  стал их делать константами, кроме delta:)

Comment (Nikita94) [inline code/kochetov-pavel/src/test/java/ru/unn/agile/Credit/Model/CreditTest.java]: Достаточно оформить delta как константу и все. Другое пусть будет так.

**Pull request merged**